File: frontend/src/api/finance.ts
/** 合同应收、收款、核销和退款闭环 API。 */
import request from './request'

export interface Receivable {
  id: number
  contract_id: number
  sequence_no: number
  title: string
  amount: number | string
  due_date: string
  currency: string
  status: string
  allocated_amount: number | string
  outstanding_amount: number | string
  effective_status: string
  contract_no?: string | null
  contract_title?: string | null
  customer_name?: string | null
  owner_name?: string | null
  version: number
  remark?: string | null
  created_at: string
  updated_at: string
}

export interface Receipt {
  id: number
  receipt_no: string
  contract_id: number
  amount: number | string
  paid_date: string
  payer_name: string
  payment_method?: string | null
  bank_reference?: string | null
  proof_filename?: string | null
  status: string
  submitted_by?: number | null
  confirmed_by?: number | null
  confirmed_at?: string | null
  allocated_amount: number | string
  pending_allocation_amount?: number | string
  refunded_amount: number | string
  available_amount: number | string
  contract_no?: string | null
  contract_title?: string | null
  customer_name?: string | null
  submitted_by_name?: string | null
  confirmed_by_name?: string | null
  version: number
  remark?: string | null
  created_at: string
  updated_at: string
}

export interface ReceiptAllocation {
  id: number
  receipt_id: number
  receivable_plan_id: number
  amount: number | string
  status: 'pending' | 'active' | 'rejected' | 'reversed'
  allocated_by?: number | null
  allocated_at: string
  approved_by?: number | null
  approved_at?: string | null
  review_remark?: string | null
  reversed_by?: number | null
  reversed_at?: string | null
  reverse_reason?: string | null
  idempotency_key?: string | null
  version: number
}

export interface Refund {
  id: number
  refund_no: string
  receipt_id: number
  amount: number | string
  reason: string
  review_remark?: string | null
  status: 'pending' | 'confirmed' | 'rejected' | 'cancelled'
  version: number
  created_at: string
  updated_at: string
}

export interface ContractFinancialSummary {
  contract_id: number
  contract_amount: number | string
  receivable_total: number | string
  confirmed_receipt_total: number | string
  refunded_total: number | string
  allocated_total: number | string
  outstanding_receivable: number | string
  unallocated_receipt_balance: number | string
  overdue_receivable: number | string
}

export interface FinanceStats {
  month_contract_amount: number | string
  confirmed_receipt_amount: number | string
  receivable_total: number | string
  outstanding_receivable_amount: number | string
  allocated_amount: number | string
  unallocated_receipt_amount: number | string
  pending_review_count: number
  pending_review_amount: number | string
  overdue_count: number
  overdue_amount: number | string
  collection_rate: number | string
  forecast_gross_margin: number | string
}

export interface FinanceList<T> {
  total: number
  items: T[]
}

export function fetchReceivables(contractId: number) {
  return request.get<Receivable[]>(`/contracts/${contractId}/receivables`)
}

export function fetchReceivableWorkbench(params: {
  status?: string
  keyword?: string
  page?: number
  page_size?: number
}) {
  return request.get<FinanceList<Receivable>>('/receivables', { params })
}

export function createReceivable(
  contractId: number,
  data: {
    title: string
    amount: number
    due_date: string
    sequence_no?: number
    remark?: string
  },
) {
  return request.post<Receivable>(`/contracts/${contractId}/receivables`, data)
}

export function updateReceivable(
  id: number,
  data: {
    title?: string
    amount?: number
    due_date?: string
    remark?: string
    version: number
  },
) {
  return request.patch<Receivable>(`/receivables/${id}`, data)
}

export function cancelReceivable(id: number, version: number, reason: string) {
  return request.post<Receivable>(`/receivables/${id}/cancel`, { version, reason })
}

export function fetchReceipts(contractId: number) {
  return request.get<FinanceList<Receipt>>('/receipts', {
    params: { contract_id: contractId, page: 1, page_size: 100 },
  })
}

export function fetchReceiptWorkbench(params: {
  status?: string
  keyword?: string
  page?: number
  page_size?: number
}) {
  return request.get<FinanceList<Receipt>>('/receipts', { params })
}

export function createReceipt(data: {
  contract_id: number
  amount: number
  paid_date: string
  payer_name: string
  payment_method?: string
  bank_reference?: string
  proof_filename?: string
  idempotency_key?: string
  remark?: string
}) {
  return request.post<Receipt>('/receipts', data)
}

export function reviewReceipt(id: number, approve: boolean, version: number, remark?: string) {
  return request.post<Receipt>(`/receipts/${id}/${approve ? 'confirm' : 'reject'}`, {
    version,
    remark,
  })
}

export function fetchAllocations(receiptId: number) {
  return request.get<ReceiptAllocation[]>(`/receipts/${receiptId}/allocations`)
}

export function createAllocation(
  receiptId: number,
  data: { receivable_plan_id: number; amount: number; idempotency_key?: string },
) {
  return request.post<ReceiptAllocation>(`/receipts/${receiptId}/allocations`, data)
}

export function reviewAllocation(id: number, approve: boolean, version: number, remark?: string) {
  return request.post<ReceiptAllocation>(`/allocations/${id}/${approve ? 'confirm' : 'reject'}`, {
    version,
    remark,
  })
}

export function reverseAllocation(id: number, version: number, reason: string) {
  return request.post<ReceiptAllocation>(`/allocations/${id}/reverse`, { version, reason })
}

export function fetchRefunds(receiptId: number) {
  return request.get<Refund[]>(`/receipts/${receiptId}/refunds`)
}

export function createRefund(
  receiptId: number,
  data: { amount: number; reason: string; idempotency_key?: string },
) {
  return request.post<Refund>(`/receipts/${receiptId}/refunds`, data)
}

export function reviewRefund(id: number, approve: boolean, version: number, remark?: string) {
  return request.post<Refund>(`/refunds/${id}/${approve ? 'confirm' : 'reject'}`, {
    version,
    remark,
  })
}

export function fetchContractFinancialSummary(contractId: number) {
  return request.get<ContractFinancialSummary>(
    `/contracts/${contractId}/financial-summary`,
  )
}

export function fetchFinanceStats() {
  return request.get<FinanceStats>('/finance/stats')
}

File: backend/app/schemas/finance.py
"""合同财务闭环 API Schema。"""
from datetime import date, datetime
from decimal import Decimal
from typing import Optional

from pydantic import BaseModel, ConfigDict, Field


class ReceivableCreate(BaseModel):
    title: str = Field(..., min_length=1, max_length=100)
    amount: Decimal = Field(..., gt=0)
    due_date: date
    sequence_no: Optional[int] = Field(None, ge=1)
    remark: Optional[str] = None


class ReceivableUpdate(BaseModel):
    title: Optional[str] = Field(None, min_length=1, max_length=100)
    amount: Optional[Decimal] = Field(None, gt=0)
    due_date: Optional[date] = None
    remark: Optional[str] = None
    version: int = Field(..., ge=1)


class ReceivableCancelRequest(BaseModel):
    reason: str = Field(..., min_length=1, max_length=500)
    version: int = Field(..., ge=1)


class ReceivableOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: int
    contract_id: int
    sequence_no: int
    title: str
    amount: Decimal
    due_date: date
    currency: str
    status: str
    allocated_amount: Decimal = Decimal("0")
    outstanding_amount: Decimal = Decimal("0")
    effective_status: str = "unpaid"
    contract_no: Optional[str] = None
    contract_title: Optional[str] = None
    customer_name: Optional[str] = None
    owner_name: Optional[str] = None
    version: int
    remark: Optional[str] = None
    created_at: datetime
    updated_at: datetime


class ReceivableListOut(BaseModel):
    total: int
    items: list[ReceivableOut]


class ReceiptCreate(BaseModel):
    contract_id: int
    amount: Decimal = Field(..., gt=0)
    paid_date: date
    payer_name: str = Field(..., min_length=1, max_length=200)
    payment_method: Optional[str] = Field(None, max_length=50)
    bank_reference: Optional[str] = Field(None, max_length=100)
    proof_filename: Optional[str] = Field(None, max_length=255)
    idempotency_key: Optional[str] = Field(None, min_length=8, max_length=100)
    remark: Optional[str] = None


class ReceiptReviewRequest(BaseModel):
    remark: Optional[str] = None
    version: int = Field(..., ge=1)


class ReceiptOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: int
    receipt_no: str
    contract_id: int
    amount: Decimal
    paid_date: date
    payer_name: str
    payment_method: Optional[str] = None
    bank_reference: Optional[str] = None
    proof_filename: Optional[str] = None
    status: str
    submitted_by: Optional[int] = None
    confirmed_by: Optional[int] = None
    confirmed_at: Optional[datetime] = None
    allocated_amount: Decimal = Decimal("0")
    pending_allocation_amount: Decimal = Decimal("0")
    refunded_amount: Decimal = Decimal("0")
    available_amount: Decimal = Decimal("0")
    contract_no: Optional[str] = None
    contract_title: Optional[str] = None
    customer_name: Optional[str] = None
    submitted_by_name: Optional[str] = None
    confirmed_by_name: Optional[str] = None
    version: int
    remark: Optional[str] = None
    created_at: datetime
    updated_at: datetime


class ReceiptListOut(BaseModel):
    total: int
    items: list[ReceiptOut]


class AllocationCreate(BaseModel):
    receivable_plan_id: int
    amount: Decimal = Field(..., gt=0)
    idempotency_key: Optional[str] = Field(None, min_length=8, max_length=100)


class AllocationReverseRequest(BaseModel):
    reason: str = Field(..., min_length=1, max_length=500)
    version: int = Field(..., ge=1)


class AllocationReviewRequest(BaseModel):
    remark: Optional[str] = None
    version: int = Field(..., ge=1)


class AllocationOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: int
    receipt_id: int
    receivable_plan_id: int
    amount: Decimal
    status: str
    allocated_by: Optional[int] = None
    allocated_at: datetime
    approved_by: Optional[int] = None
    approved_at: Optional[datetime] = None
    review_remark: Optional[str] = None
    reversed_by: Optional[int] = None
    reversed_at: Optional[datetime] = None
    reverse_reason: Optional[str] = None
    idempotency_key: Optional[str] = None
    version: int


class RefundCreate(BaseModel):
    amount: Decimal = Field(..., gt=0)
    reason: str = Field(..., min_length=1, max_length=500)
    idempotency_key: Optional[str] = Field(None, min_length=8, max_length=100)


class RefundReviewRequest(BaseModel):
    remark: Optional[str] = None
    version: int = Field(..., ge=1)


class RefundOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: int
    refund_no: str
    receipt_id: int
    amount: Decimal
    reason: str
    review_remark: Optional[str] = None
    status: str
    requested_by: Optional[int] = None
    confirmed_by: Optional[int] = None
    confirmed_at: Optional[datetime] = None
    idempotency_key: Optional[str] = None
    version: int
    created_at: datetime
    updated_at: datetime


class ContractFinancialSummary(BaseModel):
    contract_id: int
    contract_amount: Decimal
    receivable_total: Decimal
    confirmed_receipt_total: Decimal
    refunded_total: Decimal
    allocated_total: Decimal
    outstanding_receivable: Decimal
    unallocated_receipt_balance: Decimal
    overdue_receivable: Decimal


class FinanceStatsOut(BaseModel):
    month_contract_amount: Decimal
    confirmed_receipt_amount: Decimal
    receivable_total: Decimal
    outstanding_receivable_amount: Decimal
    allocated_amount: Decimal
    unallocated_receipt_amount: Decimal
    pending_review_count: int
    pending_review_amount: Decimal
    overdue_count: int
    overdue_amount: Decimal
    collection_rate: Decimal
    forecast_gross_margin: Decimal = Decimal("34.8")
